browser/src/controller.js

```diff
@@ -0,0 +1,171 @@
+    canvasElement.onwheel = event => event.preventDefault();
+    canvasElement.onmousewheel = event => event.preventDefault();
+
+    document.addEventListener('keydown', event => {
+      if (event.key === ' ') {
+        event.preventDefault();
+      } else if (event.key === 'ArrowUp') {
+        event.preventDefault();
+      } else if (event.key === 'ArrowDown') {
+        event.preventDefault();
+      }
+    });
+  }
+
+  /**
+   * Adds bindings to the window
+   */
+  addWindowBindings() {
+    // TODO: why is this bound to the document instead of the window
+    document.addEventListener('mouseup', e => this.service.send(e));
+
+    // resize the canvas every time the window is resized
+    window.addEventListener('resize', () => {
+      waitForFinalEvent(() => {
+        this.view.setCanvasDimensions();
+        this.view.displayUndoRedo();
+        this.view.canvasView.render();
+      }, 500, 'canvasResize');
+    });
+
+    window.addEventListener('keydown', event => {
+      this.handleKeydown(event);
+    }, false);
+
+    window.addEventListener('keyup', event => {
+      this.handleKeyup(event);
+    }, false);
+  }
+
+  /**
+   * Adds bindings to the interactive canvas
+   */
+  addCanvasBindings() {
+    const canvasElement = document.getElementById('canvas');
+    canvasElement.addEventListener('click', e => this.service.send(e));
+    canvasElement.addEventListener('mousedown', e => this.service.send(e));
+    canvasElement.addEventListener('mousemove', e => this.service.send(e));
+
+    canvasElement.addEventListener('wheel', e => this.handleScroll(e));
+    canvasElement.addEventListener('contextmenu', e => e.preventDefault());
+    canvasElement.addEventListener('selectstart', e => e.preventDefault());
+  }
+
+  addUndoBindings() {
+    document.getElementById('undo').onclick = () => this.undo();
+    document.getElementById('redo').onclick = () => this.redo();
+  }
+
+  /**
+   * Handle mouse scroll to adjust contrast, brightness, or zoom
+   * @param {WheelEvent} event
+   */
+  handleScroll(event) {
+    if (event.altKey) {
+      this.service.send({ type: 'ZOOM', change: Math.sign(event.deltaY) })
+    } else if (event.shiftKey) {
+      // shift + scroll causes horizontal scroll on mice wheels, but not trackpads
+      const change = event.deltaY === 0 ? event.deltaX : event.deltaY;
+      this.service.send({ type: 'SCROLLBRIGHTNESS', change: change })
+    } else {
+      this.service.send({ type: 'SCROLLCONTRAST', change: event.deltaY })
+    }
+  }
+
+  /**
+   * Handle all keybinds
+   * @param {KeyboardEvent} event
+   */
+  handleKeydown(event) {
+    if ((event.ctrlKey || event.metaKey) && event.shiftKey && (event.key === 'Z' || event.key === 'z')) {
+      this.redo();
+    } else if ((event.ctrlKey || event.metaKey) && (event.key === 'Z' || event.key === 'z')) {
+      this.undo();
+    } else if ((event.ctrlKey || event.metaKey) && event.key === 'b') {
+      this.service.send('TOGGLERGB');
```

Comment:
These cases are not in the keydown lookup table because they are checking for multiple keys at once (Ctrl or Command + something else) and they also behave differently, calling a function or sending a named event instead of sending a `keydown.[key]` event. I don't see how we can fold them into the keydown lookup table. Maybe we could have function for each case, but that doesn't sound any better than what we have.